File: nudgly/services/llm_service.py
import json

from PySide6.QtCore import QObject, Signal, Slot, QBuffer, QIODevice, QByteArray
from PySide6.QtGui import QGuiApplication, QPixmap
from PySide6.QtNetwork import QNetworkAccessManager, QNetworkRequest, QNetworkReply

from nudgly.constants import Constants
from nudgly.models.conversation_model import Message, TextPart, ImageUrlPart, ConversationModel
from nudgly.services.config_service import ConfigService
from nudgly.services.logging_service import LoggingService


class LLMService(QObject):
    answerGenerated = Signal(str)
    errorOccurred = Signal(str)

    def __init__(self, parent=None):
        super().__init__(parent)
        self._API_KEY = ConfigService.read_value("LLM", "api_key")
        self._network_manager = QNetworkAccessManager(self)

    @staticmethod
    def _to_base64_png(pixmap: QPixmap) -> str | None:
        if pixmap is None:
            return None

        buffer = QBuffer()
        buffer.open(QIODevice.OpenModeFlag.WriteOnly)

        image = pixmap.toImage()
        image.save(buffer, "PNG")

        byte_array = buffer.data()
        buffer.close()

        return byte_array.toBase64().data().decode("utf-8")

    @staticmethod
    def take_screenshot() -> str | None:
        screen = QGuiApplication.primaryScreen()

        if not screen:
            LoggingService.error("Unable to get primary screen.")
            return None

        pixmap = screen.grabWindow(0)

        return LLMService._to_base64_png(pixmap)

    @Slot(ConversationModel)
    def generate_answer(self, conversation_model: ConversationModel):
        base64_encoded_screenshot = self.take_screenshot()

        if not base64_encoded_screenshot:
            error_msg = "Capturing screen failed."
            LoggingService.error(error_msg)
            self.errorOccurred.emit(error_msg)
            return

        user_text_part: TextPart = {
            "type": "text",
            "text": Constants.DEFAULT_SYSTEM_PROMPT,
        }
        user_image_part: ImageUrlPart = {
            "type": "image_url",
            "image_url": {"url": f"data:image/png;base64,{base64_encoded_screenshot}"}
        }
        user_message: Message = {
            "role": "user",
            "content": [user_text_part, user_image_part]
        }

        self._send_api_request(user_message, conversation_model)

    def _send_api_request(self, user_message: Message, conversation_model: ConversationModel):
        conversation_data = conversation_model.get_conversation_data()
        conversation_data["messages"].append(user_message)

        request = QNetworkRequest(Constants.MODEL_ENDPOINT)
        request.setRawHeader(b"Authorization", f"Bearer {self._API_KEY}".encode())
        request.setRawHeader(b"Content-Type", "application/json".encode())

        json_payload = json.dumps(conversation_data).encode('utf-8')
        data = QByteArray(json_payload)

        LoggingService.info("Sending request to Gemini API...")
        reply = self._network_manager.post(request, data)

        # Connect the reply's finished signal to a handler slot.
        # Use a lambda to pass the reply and conversation model to the handler.
        reply.finished.connect(lambda: self._handle_api_response(reply, conversation_model, user_message))

    def _handle_api_response(self, reply: QNetworkReply, conversation_model: ConversationModel,
                             user_message: Message) -> None:
        if reply.error() != QNetworkReply.NetworkError.NoError:
            error_string = reply.errorString()
            response_body = reply.readAll().data().decode('utf-8')
            LoggingService.error(f"Error making API request: {error_string}")
            LoggingService.info(f"Response Body: {response_body}")
            self.errorOccurred.emit(error_string)
        else:
            LoggingService.info("Request successful.")
            response_data = reply.readAll().data()
            try:
                output = json.loads(response_data)

                conversation_model.add_message(user_message["role"], user_message["content"])

                assistant_text = output["choices"][0]["message"]["content"]
                assistant_message: Message = {
                    "role": output["choices"][0]["message"]["role"],
                    "content": [{"type": "text", "text": assistant_text}]
                }
                conversation_model.add_message(assistant_message["role"], assistant_message["content"])

                self.answerGenerated.emit(assistant_text)

            except (json.JSONDecodeError, KeyError) as e:
                error_msg = f"Error parsing API response: {e}"
                LoggingService.error(error_msg)
                self.errorOccurred.emit(error_msg)

        reply.deleteLater()

File: src/Services/LLMService.cpp
#include "Services/LLMService.h"

#include <QScreen>
#include <QBuffer>
#include <QNetworkAccessManager>
#include <QNetworkRequest>
#include <QNetworkReply>
#include <QJsonObject>
#include <QJsonArray>
#include <QVariantMap>
#include <QPixMap>
#include <QtLogging>

#include "Constants.h"
#include "Services/SettingsService.h"


// --- JSON Serialization Helpers ---
namespace
{
    QJsonObject toJson(const Models::ConversationModel::ImageUrl& imageUrl)
    {
        return {{"url", imageUrl.url}};
    }

    QJsonObject toJson(const Models::ConversationModel::TextPart& textPart)
    {
        return {{"type", "text"}, {"text", textPart.text}};
    }

    QJsonObject toJson(const Models::ConversationModel::ImageUrlPart& imagePart)
    {
        return {{"type", "image_url"}, {"image_url", toJson(imagePart.imageUrl)}};
    }

    QJsonArray toJson(const QList<Models::ConversationModel::ContentPart>& contentList)
    {
        QJsonArray contentArray;
        for (const auto& partVariant : contentList)
        {
            if (partVariant.canConvert<Models::ConversationModel::TextPart>())
            {
                contentArray.append(toJson(partVariant.value<Models::ConversationModel::TextPart>()));
            }
            else if (partVariant.canConvert<Models::ConversationModel::ImageUrlPart>())
            {
                contentArray.append(toJson(partVariant.value<Models::ConversationModel::ImageUrlPart>()));
            }
        }
        return contentArray;
    }
} // end anonymous namespace

Services::LLMService::LLMService(QObject* parent)
    : QObject(parent),
      m_networkManager(new QNetworkAccessManager(this))
{
    m_apiKey = SettingsService::readValue("LLM/apiKey").toString();
}

QString Services::LLMService::toBase64Png(const QPixmap& pixmap)
{
    QByteArray byteArray;
    QBuffer buffer(&byteArray);
    buffer.open(QIODevice::WriteOnly);
    pixmap.save(&buffer, "PNG");
    return QString::fromUtf8(byteArray.toBase64());
}

QString Services::LLMService::takeScreenshot()
{
    QScreen* screen = QGuiApplication::primaryScreen();
    if (!screen)
    {
        qCritical() << "Unable to get primary screen.";
        return {};
    }
    const QPixmap pixmap = screen->grabWindow(0);
    return toBase64Png(pixmap);
}

void Services::LLMService::generateAnswer(Models::ConversationModel::ConversationModel* conversationModel)
{
    const QString base64Screenshot = takeScreenshot();
    if (base64Screenshot.isEmpty())
    {
        const QString errorMsg = "Capturing screen failed.";
        qCritical() << errorMsg;
        emit errorOccurred(errorMsg);
        return;
    }

    Models::ConversationModel::TextPart textPart;
    textPart.text = Constants::DEFAULT_PROMPT;

    Models::ConversationModel::ImageUrlPart imagePart;
    imagePart.imageUrl.url = QString("data:image/png;base64,%1").arg(base64Screenshot);

    Models::ConversationModel::Message userMessage;
    userMessage.role = "user";
    userMessage.content.append(QVariant::fromValue(textPart));
    userMessage.content.append(QVariant::fromValue(imagePart));

    conversationModel->addMessage(userMessage);

    Models::ConversationModel::Message assistantPlaceholder;
    assistantPlaceholder.role = "assistant";
    conversationModel->addMessage(assistantPlaceholder);

    sendApiRequest(conversationModel);
}

void Services::LLMService::sendApiRequest(
    Models::ConversationModel::ConversationModel* conversationModel)
{
    auto [model, messages] = conversationModel->getConversationData();
    // messages.append(userMessage);

    // Create the final JSON payload from the Conversation object
    QJsonObject payloadObject;
    payloadObject["model"] = model;
    QJsonArray messagesArray;
    for (const auto& [role, content] : messages)
    {
        QJsonObject msgObject;
        msgObject["role"] = role;
        msgObject["content"] = toJson(content);
        messagesArray.append(msgObject);
    }
    payloadObject["messages"] = messagesArray;
    payloadObject["stream"] = true;

    QNetworkRequest request(Constants::MODEL_ENDPOINT);
    request.setRawHeader("Authorization", ("Bearer " + m_apiKey).toUtf8());
    request.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");

    m_fullResponse.clear();
    m_streamBuffer.clear();

    m_currentReply = m_networkManager->post(request, QJsonDocument(payloadObject).toJson());
    m_currentModel = conversationModel; // Store model for use in slots

    // Connect to the readyRead signal to process chunks as they arrive
    connect(m_currentReply, &QNetworkReply::readyRead, this, &LLMService::handleApiStream);

    // Connect to the finished signal for cleanup and finalization
    connect(m_currentReply, &QNetworkReply::finished, this, &LLMService::handleApiFinished);

    // Connect to the error signal for robust error handling
    connect(m_currentReply, &QNetworkReply::errorOccurred, this, [this](const QNetworkReply::NetworkError code)
    {
        qCritical() << "Network error occurred:" << code << m_currentReply->errorString();
        emit errorOccurred(m_currentReply->errorString());
    });
}

void Services::LLMService::handleApiStream()
{
    // Append new data from the reply to our buffer
    m_streamBuffer.append(m_currentReply->readAll());

    // Process the buffer line by line. SSE events are separated by "\n\n".
    QList<QByteArray> events = m_streamBuffer.split('\n');
    m_streamBuffer = events.takeLast(); // Keep incomplete line in buffer

    for (const QByteArray& event : events)
    {
        if (event.startsWith("data: "))
        {
            QByteArray jsonData = event.mid(6);
            if (jsonData.trimmed() == "[DONE]")
            {
                continue; // The 'finished' signal will handle the end
            }

            QJsonDocument doc = QJsonDocument::fromJson(jsonData);
            if (!doc.isObject()) continue;

            // Extract the text chunk from the delta
            const QString chunk = doc.object()["choices"].toArray()[0]
                .toObject()["delta"]
                .toObject()["content"].toString();

            if (!chunk.isEmpty())
            {
                m_fullResponse.append(chunk);
                m_currentModel->appendToLastMessage(chunk);
                emit answerChunkReceived(chunk);
            }
        }
    }
}

void Services::LLMService::handleApiFinished()
{
    if (m_currentReply->error() == QNetworkReply::NoError)
    {
        // Handle any remaining data in the buffer
        handleApiStream();
        qInfo() << "API stream finished successfully.";
        emit answerGenerated(m_fullResponse);
    }
    else
    {
        // Error signal will have already been emitted.
        qCritical() << "API stream finished with error:" << m_currentReply->errorString();
    }

    m_currentReply->deleteLater();
    m_currentReply = nullptr;
    m_currentModel = nullptr;
}
